kv/kvserver: TestRestoreReplicas failed

kv/kvserver.TestRestoreReplicas [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10064002?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10064002?buildTab=artifacts#/) on master @ [63eb6b86c1518fbed9db10e7d2afa1446b2b3b82](https://github.com/cockroachdb/cockroach/commits/63eb6b86c1518fbed9db10e7d2afa1446b2b3b82):


```
=== RUN   TestRestoreReplicas
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestRestoreReplicas3865623463
    test_log_scope.go:79: use -show-logs to present logs inline
    client_raft_test.go:402: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:402
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 1
        	Test:       	TestRestoreReplicas
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestRestoreReplicas3865623463
--- FAIL: TestRestoreReplicas (12.31s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27902

Epic CRDB-27234

kv/kvserver.TestRestoreReplicas [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10074597?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10074597?buildTab=artifacts#/) on master @ [63eb6b86c1518fbed9db10e7d2afa1446b2b3b82](https://github.com/cockroachdb/cockroach/commits/63eb6b86c1518fbed9db10e7d2afa1446b2b3b82):


```
=== RUN   TestRestoreReplicas
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestRestoreReplicas3808935525
    test_log_scope.go:79: use -show-logs to present logs inline
    client_raft_test.go:402: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:402
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 1
        	Test:       	TestRestoreReplicas
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestRestoreReplicas3808935525
--- FAIL: TestRestoreReplicas (4.52s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestRestoreReplicas [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10085577?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10085577?buildTab=artifacts#/) on master @ [a65bca232dfde1862e5aff343e7b1f3a379d6cd6](https://github.com/cockroachdb/cockroach/commits/a65bca232dfde1862e5aff343e7b1f3a379d6cd6):


```
=== RUN   TestRestoreReplicas
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestRestoreReplicas3103863021
    test_log_scope.go:79: use -show-logs to present logs inline
    client_raft_test.go:402: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:402
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 1
        	Test:       	TestRestoreReplicas
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestRestoreReplicas3103863021
--- FAIL: TestRestoreReplicas (5.22s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestRestoreReplicas [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10099314?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10099314?buildTab=artifacts#/) on master @ [19e57552fcfa8bb6f07945f699e67f3b8179be89](https://github.com/cockroachdb/cockroach/commits/19e57552fcfa8bb6f07945f699e67f3b8179be89):


```
=== RUN   TestRestoreReplicas
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/7fb080117a3da2d7cdf3576571d60489/logTestRestoreReplicas179368972
    test_log_scope.go:79: use -show-logs to present logs inline
    client_raft_test.go:402: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:402
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 2
        	Test:       	TestRestoreReplicas
        	Messages:   	replica command failed (non leaseholders)
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7fb080117a3da2d7cdf3576571d60489/logTestRestoreReplicas179368972
--- FAIL: TestRestoreReplicas (4.76s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -373,37 +373,39 @@ func TestRestoreReplicas(t *testing.T) {
 
 	// Perform an increment before replication to ensure that commands are not
 	// repeated on restarts.
-	incArgs := incrementArgs([]byte("a"), 23)
-	if _, err := kv.SendWrapped(ctx,
-		store.TestSender(), incArgs); err != nil {
-		t.Fatal(err)
-	}
+	incArgs := incrementArgs(key, 23)
+	_, pErr := kv.SendWrapped(ctx, store.TestSender(), incArgs)
+	require.NoError(t, pErr.GoError())
 
 	tc.AddVotersOrFatal(t, key, tc.Target(1))
 
 	require.NoError(t, tc.Restart())
 
-	incArgs = incrementArgs([]byte("a"), 5)
-	failures := 0
-	successes := 0
-	// Send a command on each store. It should only succeed on the lease holder.
+	// Find the leaseholder and follower. The restart may cause the Raft
+	// leadership to bounce around a bit, since we don't fully enable Raft
+	// prevote, so we loop for a bit until we find the leaseholder.
+	incArgs = incrementArgs(key, 5)
 	var followerStore *kvserver.Store
-	for i := 0; i < len(tc.Servers); i++ {
-		if _, pErr := kv.SendWrapped(ctx, tc.GetFirstStoreFromServer(t, i).TestSender(), incArgs); pErr != nil {
-			failures++
-			if _, ok := pErr.GetDetail().(*kvpb.NotLeaseHolderError); !ok {
-				t.Fatalf("expected not lease holder error; got %s", pErr)
+	testutils.SucceedsSoon(t, func() error {
+		var pErr *kvpb.Error
+		for i := 0; i < tc.NumServers(); i++ {
+			_, pErr = kv.SendWrapped(ctx, tc.GetFirstStoreFromServer(t, i).TestSender(), incArgs)
+			if pErr == nil {
+				followerStore = tc.GetFirstStoreFromServer(t, 1-i)
+				break
 			}
-			followerStore = tc.GetFirstStoreFromServer(t, i)
-		} else {
-			successes++
+			require.IsType(t, &kvpb.NotLeaseHolderError{}, pErr.GetDetail())
 		}
-	}
-	require.Equal(t, 1, failures, "replica command failed (non leaseholders)")
-	require.Equal(t, 1, successes, "replica command succeeded (leaseholders)")
+		return pErr.GoError()
+	})
+
+	// The follower should now return a NLHE.
+	_, pErr = kv.SendWrapped(ctx, followerStore.TestSender(), incArgs)
+	require.Error(t, pErr.GoError())
+	require.IsType(t, &kvpb.NotLeaseHolderError{}, pErr.GetDetail())
 
 	testutils.SucceedsSoon(t, func() error {
-		getArgs := getArgs([]byte("a"))
+		getArgs := getArgs(key)
 		if reply, err := kv.SendWrappedWith(ctx, followerStore.TestSender(), kvpb.Header{
 			ReadConsistency: kvpb.INCONSISTENT,
 		}, getArgs); err != nil {

